### PR TITLE
Add Idea Hub Gutenberg notices.

### PR DESCRIPTION
**Pull request by tofumatt**
## Summary

Addresses issue #3272.

## Relevant technical choices

I needed to add the ability for the Idea Hub postmeta to be exposed in the REST API; I went with the `$args` approach as it was straightforward and quick, but obviously it moves the source of truth from the meta file to the instantiation of the meta. This might not be ideal, but I wasn't sure.

Let me know what you think on that point, we can refactor it to be something fancier but I didn't want to start modifying too many classes 😅 

Regarding the IB, I put everything in one listener rather than listening for the `hasNotices` in its own listener, because we need to wait for the postmeta requests to resolve anyway, which involves a listener. I think my approach is a bit less complex than the one proposed in the IB, and it accomplishes the same thing 🙂 

It might be nice to eventually store the "has viewed" meta on the server, as we _do_ have a TTL on these and it might get annoying to see the notice in different browsers/computers, but for now it's client-side only.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Modules/Idea_Hub.php
Comment (tofumatt) [inline includes/Modules/Idea_Hub.php:124]: This is what I meant by moving the source of truth. A simple `bool` for `show_in_rest` isn't a good idea either, because complex values (like JSON arrays, etc.) require a transform to handle how they should be displayed. We could add that as some extra `Post_Meta` metadata, I suppose.

File: includes/Core/Assets/Assets.php
Comment (eugene-manuilov) [inline includes/Core/Assets/Assets.php:83]: No need to add a dummy comment for the initial version of the method. Just the version number is enough:

```suggestion
	 * @since 1.0.0
```

File: includes/Core/Modules/Module_With_Assets.php
Comment (eugene-manuilov) [inline includes/Core/Modules/Module_With_Assets.php:36]: Same here:
```suggestion
	 * @since 1.7.0
```

File: includes/Core/Modules/Module_With_Assets_Trait.php
Comment (eugene-manuilov) [inline includes/Core/Modules/Module_With_Assets_Trait.php:53]: And here:
```suggestion
	 * @since 1.7.0
```

File: includes/Core/Storage/Post_Meta_Setting.php
Comment (eugene-manuilov) [inline includes/Core/Storage/Post_Meta_Setting.php:53]: @tofumatt, although this approach works, we should better solve the problem with the `show_in_rest` option using the OOP approach. What we need to do is the following:

1. Remove the new `$args` approach.
1. Add a new protected method `get_show_in_rest` that returns `false` by default.
1. Update the `register_meta` function call to be as the following:
    ```php
    register_meta(
        'post',
        static::META_KEY,
        array(
            'type'              => $this->get_type(),
            'sanitize_callback' => $this->get_sanitize_callback(),
            'single'            => true,
            'show_in_rest'      => $this->get_show_in_rest(),
        )
    );
    ```
1. Add the new method `get_show_in_rest` to the `Post_Idea_Text` class that returns `true`.

File: includes/Modules/Idea_Hub.php
Comment (eugene-manuilov) [inline includes/Modules/Idea_Hub.php:511]: Seems like we don't need these dependencies for this script, do we? Let's remove them then.

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:58]: Let's return early if this condition hasn't met:
```suggestion
		if ( hasNotice( postID ) !== false ) {
			return;
		}
```

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:46]:
```suggestion
	const hasNotice = ( postID ) => {
		const notices = wp.data.select( 'core/notices' ).getNotices();
		if ( notices === undefined ) {
			return undefined;
		}

		const key = editorNoticeKey( postID );
		return notices.some( ( { id } ) => id === key );
	};
```

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:56]: The `editorNoticeKey( postID )` function call repeats four times. Please, assign it to a variable and use the variable instead of the same function calls.

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:29]: This variable belongs to the `loadIdeaHubNotices` function context. Let's move it into that function then.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @tofumatt. This is a good start. However, there are a few things (+ some nitpicks) that we need to address before we can merge this PR.

File: includes/Core/Storage/Post_Meta_Setting.php
Comment (tofumatt) [reply includes/Core/Storage/Post_Meta_Setting.php:53]: Right on, that's nicer 🙂

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:49]: This function is a little bit of a mess. Please, update it to have the following flow:

1. Get postID
    * If postID is falsy, return
1. Get the item from cache
    * If cacheHit is truthy, unsubscribe and return
1. Get the hasNotice result
    * If it returns `false` and `shownNotices` contains `noticeKey`, then call `setItem`, unsubscribe and return
1. Get post meta
    * If idea_text is not empty, show the notice

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @tofumatt. Added two more comments. I believe it will be good to go once we address the remaining issues.

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (tofumatt) [reply assets/js/googlesitekit-idea-hub-notice.js:49]: Thanks, that's much better! 👍🏻

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (eugene-manuilov) [inline assets/js/googlesitekit-idea-hub-notice.js:95]: @tofumatt, we need to make sure that we create a new notice only if the `shownNotices` does not include `noticeKey`.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Found one more issue. Matthew, could you please take a look?

File: assets/js/googlesitekit-idea-hub-notice.js
Comment (tofumatt) [reply assets/js/googlesitekit-idea-hub-notice.js:95]: 🤦🏻 Whoops, my bad, missed it in the refactor. Thanks for catching that!

Comment (tofumatt) [reply assets/js/googlesitekit-idea-hub-notice.js:95]: Hmm, actually, I think it'd be rare that `shownNotices.includes( noticeKey )` happens here because the above `if ( hasNotice( postID ) === false && shownNotices.includes( noticeKey ) )` will return early.

But just in case I've added ` && ! shownNotices.includes( noticeKey )` to this conditional.

**Review by eugene-manuilov — APPROVED** [review]
Awesome! Looks good to me!

**Pull request merged**
